### PR TITLE
Replace `Vec::drain` by a method that accepts a range parameter.

### PR DESCRIPTION
**Pull request by mahkoh**
[Rendered](https://github.com/mahkoh/rfcs/blob/drain_range/text/0000-drain-range.md)

### CODE REVIEW

**Comment by P1start** [comment]
I’m probably being stupid, but why can’t similar versions of these methods be implemented for `String` as well (presumably yielding `char`s)?

**Comment by mahkoh** [comment]
You can't use the same implementation as the Vec one because the chars are not stored as chars. You can already achieve this by calling `str::slice(...).chars(); str::remove_range(...)` because everything is `Copy`. `remove_range` comes from #570.

**Comment by P1start** [comment]
You might not be able to share an implementation, but surely you could still have the new & improved `drain` method on `String`, even if its implementation were different? The API would be more consistent that way, in my opinion, and `drain` with a range parameter is strictly more useful than a `remove_range` method, even on a string, where `char`s are `Copy`.

**Comment by mahkoh** [comment]
A `drain` would not provide new functionality, would be slower in unoptimized builds, and would make compile times longer. A `remove_range` function on String is orthogonal to the current set of methods.

It's unfortunate that `drain` on `Vec` will also have these problems but I don't think we can add both `drain` and `remove_range`.

**Comment by aturon** [comment]
I like the idea of this RFC as well as the [one for `String](https://github.com/rust-lang/rfcs/pull/570), but like @P1start would strongly prefer for the APIs to match.

@mahkoh, can you provide any evidence that the performance hit on debug builds is bad enough to merit API inconsistency at this stage?

**Comment by mahkoh** [comment]
I've updated the RFC. Details are left to the implementation.

**Comment by bombless** [comment]
Seems good.
Is it possible to take `Iterator` as range?
Say, we add `+` and `==` to Iterator, and make its `Output` like an old-school `struct Drain` that `impl`s `Iterator`.
I'm trying to figure out the lifetime thing for this.

**Comment by huonw** [comment]
There's no drawbacks to this design?

**Comment by mahkoh** [comment]
None. Perfect RFC.

At least none that I'm aware of. If you have some I can add them.

**Comment by huonw** [comment]
There's at least the obvious one of making things more generic and hence the documentation harder to use. It also causes the API for the `drain` method to differ between `Vec`/`String` and other types.

File: text/0000-drain-range.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+```rust
+/// Creates a draining iterator that clears the specified range in the String
+/// and iterates over the characters contained in the range.
+///
+/// # Panics
+///
+/// Panics if the range is decreasing, if the upper bound is larger than the
+/// length of the String, or if the start and the end of the range don't lie on
+/// character boundaries.
+pub fn drain(&mut self, range: /* ? */) -> /* ? */ {
```
Comment (huonw) [inline text/0000-drain-range.md]: Why are these omitted?

Comment (mahkoh) [inline text/0000-drain-range.md]: Assuming that the return value can only be constructed inside the String module, these parts of the signature are merely an implementation detail. If you want to you can assume
```
pub fn drain<T: StringDrainer>(&mut self, range: T) -> CharRangeIter<T> {
```
or something similar.

**Comment by mahkoh** [comment]
Updated with drawbacks.

File: text/0000-drain-range.md (hunk 1 above)
Comment (huonw) [inline text/0000-drain-range.md]: The public signature doesn't seem like an implementation detail to me.

**Comment by aturon** [comment]
@mahkoh Thanks for the revisions, this definitely feels better to me. Would you mind filling in the details that @huonw requested?

Also, rather than a `Drainer` trait, I'd prefer to add a general trait for abstracting over all range type. (This request has already come up in the past). You could update the RFC text to use such a trait, but it's fine if you don't spell out the full details.

Otherwise, I think this is ready to go.

**Comment by mahkoh** [comment]
I've updated the detailed design.

**Comment by aturon** [comment]
@mahkoh, thanks, I've added this to the agenda for reaching a final decision.

**Comment by aturon** [comment]
After discussion on the core team and with stakeholders here and on IRC, this RFC has been approved.

[Tracking issue](https://github.com/rust-lang/rust/issues/23055)

**Comment by Gankro** [comment]
drain_range is apparently unsound due to safe code apparently not being able to guarantee `drop` being called. That is `mem::forget` is strictly speaking safe to implement and use.

**Pull request merged**